CIP-0101 | Integration of `keccak256` into Plutus (#524)
cardano-foundation/CIPs: CIP-????/README.md

---
CIP: ?
Title: Integration of `keccak256` into Plutus
Status: Proposed
Category: Plutus
Authors: 
  - Sergei Patrikeev <so.patrikeev@gmail.com>
  - Iñigo Querejeta-Azurmendi <inigo.querejeta@iohk.io>
Implementors: []
Discussions: 
  - https://github.com/cardano-foundation/CIPs/pull/524
Created: 2023-06-13
License: Apache-2.0
---

## Abstract
This CIP proposes an extension of the current Plutus functions to provide support for the [`keccak256`](https://keccak.team/files/Keccak-submission-3.pdf) hashing algorithm,
primarily to ensure compatibility with Ethereum's cryptographic infrastructure.

> ## Motivation

Reviewer: ```suggestion
## Motivation: why is this CIP necessary?
```
This is a standard title which helps readers & implementors also understand what's being answered here (even if authors understand it) 😎

## Specification
This proposal aims to introduce a new built-in hash function `keccak_256`.

This function will be developed following the [`keccak256`](https://keccak.team/files/Keccak-submission-3.pdf) specification 
and will utilize the [cryptonite](https://github.com/haskell-crypto/cryptonite/blob/master/Crypto/Hash/Keccak.hs) implementation. 
Since `cryptonite` is already a part of the [`cardano-base`](https://github.com/input-output-hk/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Crypto/Hash/Keccak256.hs), 
this simplifies its integration into Plutus. The cost of the `keccak_256` operation will scale linearly with the length of the message.

More specifically, Plutus will gain the following primitive operation:

* `keccak_256` :: ByteString -> ByteString

The input to this function can be a `ByteString` of arbitrary size, and the output will be a `ByteString` of 32 bytes. 
Note that this function aligns with the format of existing hash functions in Plutus, such as [blake2b_256](https://github.com/input-output-hk/plutus/blob/75267027f157f1312964e7126280920d1245c52d/plutus-core/plutus-core/src/Data/ByteString/Hash.hs#L25)

> ## Rationale

Reviewer: ```suggestion
## Rationale: how does this CIP achieve its goals?
```
(likewise with Motivation)

While the `keccak256` function might be implemented in on-chain scripts, doing so would be computationally unfeasible. 

### Trustworthiness of Implementations
The implementation of `keccak256` is based on the version found in the [cardano-base](https://github.com/input-output-hk/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Crypto/Hash/Keccak256.hs) library.

> ## Backward Compatibility

Reviewer: ```suggestion
### Backward Compatibility
```
There is no top-level section by this name (anymore) so demoting the heading will absorb this into `Rationale` which generally makes sense.  It might also make sense to drop this section there there are no backward compatibility issues here.

The addition of `keccak256` is backward compatible and will not interfere with existing functions within Plutus.

## Path to Active

> Upon successful review and evaluation, the `keccak256` function is planned for release in an upcoming update.

Reviewer: We are looking for a section here which discriminates between Acceptance and Implementation: https://github.com/cardano-foundation/CIPs/blob/master/.github/CIP-TEMPLATE.md#path-to-active

Sometimes these are trivial, e.g. "Acceptance is determined when the Plutus team(s) agree to do it" but please add whatever you can now about what actually needs to happen and who will be responsible for it.

When this comes up at the CIP editors' meeting (please attend if you can) I'll plan to mention this question along with the question of whether the overall level of detail is sufficient.

Reviewer: p.s. @serejke I can see that an expansion of this section was underway in https://github.com/cardano-foundation/CIPs/pull/528 so we'll look forward to reviewing those additions here in this PR branch. 🤓

Author: @rphair I have a multi-stage approval: first, in my fork by the more experienced colleague @iquerejeta :-), then pushing into this branch. We're going to update the PR very soon. Thanks for keeping an eye on it 🙏

Author: Done! Look forward to your opinion

> ## References

Reviewer: There is also no `References` section in our standard template: https://github.com/cardano-foundation/CIPs/blob/master/.github/CIP-TEMPLATE.md

Depending on the Markdown format you use, these might remain at the end of your document as footnotes.  Otherwise the reference links should be included inline within the document where they are referenced.

- [CIP-49](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0049/README.md): ECDSA and Schnorr signatures in Plutus Core 
- keccak256: https://keccak.team/files/Keccak-submission-3.pdf
- Ethereum Yellow Paper: https://ethereum.github.io/yellowpaper/paper.pdf

> - Ethereum standard on signing data: https://eips.ethereum.org/EIPS/eip-712

Reviewer: ```suggestion
- Ethereum standard on signing data: https://eips.ethereum.org/EIPS/eip-712

## Copyright
This CIP is licensed under [Apache-2.0][https://www.apache.org/licenses/LICENSE-2.0]. 
```
Finally our template includes a Copyright section at the end matches the one named in the document header: https://github.com/cardano-foundation/CIPs/blob/master/CIP-0001/README.md#licensing

---
CIP: ?
Title: Integration of `keccak256` into Plutus
Status: Proposed
Category: Plutus
Authors: 
  - Sergei Patrikeev <so.patrikeev@gmail.com>
  - Iñigo Querejeta-Azurmendi <inigo.querejeta@iohk.io>
Implementors: []
Discussions: 
  - https://github.com/cardano-foundation/CIPs/pull/524
Created: 2023-06-13
License: Apache-2.0
---

## Abstract
This CIP proposes an extension of the current Plutus functions to provide support for the [`keccak256`](https://keccak.team/files/Keccak-submission-3.pdf) hashing algorithm,
primarily to ensure compatibility with Ethereum's cryptographic infrastructure.

> ## Motivation: why is this CIP necessary?

Reviewer 2: Sounds good. @hai-kreate, @serejke , could you add a brief description of the use cases you had in mind? A brief description of the bridge or whatever use case that requires compatibility with Ethereum.

> ## Motivation: why is this CIP necessary?

Author: I have added the use case we're actively work on ([Hyperlane](https://hyperlane.xyz/)). Hyperlane is a growing interchain communication protocol and we've figured out a way to _partially_ support Cardano by using its native crypto hash function (`blake2b`). Having the `keccak256` would provide additional guarantees (censorship resistance) and allow for cross-chain slashing (work in progress).

> ## Motivation: why is this CIP necessary?

Reviewer: @michaelpj looks like fixed by https://github.com/cardano-foundation/CIPs/pull/524/commits/86ac8d20b7b72da354c022d726e33476eab758f2; please unresolve if not?

> ## Motivation: why is this CIP necessary?

Reviewer 3: If it helps, I can propose a well-scoped use case that can be a baseline test. The TLDR is to post verifiable messages signed from the EVM world to Cardano. The go-to signature scheme over there is `ecdsa-secp256k1` which signs on fixed-size messages. It's then natural to sign on message hash to support arbitrary message size, and the go-to hash function over there is `keccak256`. 
1. With `keccak256` in Plutus, a minting policy issuing authentic tokens can hash the posted message (to an output datum) with `keccak256` then verify it against the signature posted in redeemer.
2. Without it, the minting policy would have to trust the messenger, or force EVM agents to sign on `blake2b` hashes or sign with `ed25519`, etc.

We're asking signers to sign on both `keccak256` and `blake2b` hashes at the moment for our project, but that wouldn't scale to a larger EVM ecosystem. Adding `keccak256` to Plutus, on the other hand, would open many doors for future integrations, like as @serejke outlined with Hyperlane.

I'm more than happy to work on this test case if accepted. We already have it written in Helios.

The [integration](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0049/README.md) of the ECDSA and Schnorr signatures over the secp256k1 curve into Plutus was a significant 
step towards interoperability with Ethereum and Bitcoin ecosystems. However, full compatibility is still impossible 
due to the absence of the `keccak256` hashing algorithm in Plutus interpreter, 
which is a fundamental component of Ethereum's cryptographic framework:
- data signing standard [EIP-712](https://eips.ethereum.org/EIPS/eip-712), 
- `keccak256` is the hashing algorithm underlying Ethereum's ECDSA signatures. 
- EVM heavily [depends](https://ethereum.github.io/yellowpaper/paper.pdf) on `keccak256` for internal state management

Adding `keccak256` to Plutus would enhance the potential for cross-chain solutions between Cardano and EVM-based blockchains.

## Specification
This proposal aims to introduce a new built-in hash function `keccak_256`.

This function will be developed following the [`keccak256`](https://keccak.team/files/Keccak-submission-3.pdf) specification 

> and will utilize the [cryptonite](https://github.com/haskell-crypto/cryptonite/blob/master/Crypto/Hash/Keccak.hs) implementation. 

Reviewer 2: We do have that as a low priority task. The reason why we want to move away from it is because we have cryptonite as a dependency for a single primitive (SHA3), not due to any security concerns. However, we are not planning on moving away from cryptonite anytime soon, so using it's keccak implementation is the easiest. 

If we ever make this change in the near future (moving away from cryptonite) there will be no effects on the interface or behaviour of the primitives, so there would not need to be a hard-fork.

> and will utilize the [cryptonite](https://github.com/haskell-crypto/cryptonite/blob/master/Crypto/Hash/Keccak.hs) implementation. 

Reviewer: noted... resolving these 2 `cryptonite` conversations to highlight practical blocks to merging the CIP.

> Since `cryptonite` is already a part of the [`cardano-base`](https://github.com/input-output-hk/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Crypto/Hash/Keccak256.hs), 

Reviewer: noted... resolving these 2 `cryptonite` conversations (answered in https://github.com/cardano-foundation/CIPs/pull/524#discussion_r1252131805) to highlight practical blocks to merging the CIP.

this simplifies its integration into Plutus. The cost of the `keccak_256` operation will scale linearly with the length of the message.

More specifically, Plutus will gain the following primitive operation:

* `keccak_256` :: ByteString -> ByteString

The input to this function can be a `ByteString` of arbitrary size, and the output will be a `ByteString` of 32 bytes. 
Note that this function aligns with the format of existing hash functions in Plutus, such as [blake2b_256](https://github.com/input-output-hk/plutus/blob/75267027f157f1312964e7126280920d1245c52d/plutus-core/plutus-core/src/Data/ByteString/Hash.hs#L25)

## Rationale: how does this CIP achieve its goals?
While the `keccak256` function might be implemented in on-chain scripts, doing so would be computationally unfeasible. 

## Path to Active
This CIP may transition to active status once the Plutus version containing the `keccak_256` function is introduced 

> in a node release and becomes available on the main-net.

Reviewer: ```suggestion
in a node release and becomes available on the Mainnet.
```
also: the word "the" is optional here (not needed and usually not used)

### Acceptance Criteria
* A Plutus binding is created for the `keccak256` function and included in a new version of Plutus.
* Integration tests, similar to those of the existing Plutus hash functions, are added to the testing infrastructure.
* The function is benchmarked to assess its cost.

> * E2E tests with real use cases are incorporated into the testing infrastructure.

Reviewer: @michaelpj looks like fixed by https://github.com/cardano-foundation/CIPs/pull/524/commits/86ac8d20b7b72da354c022d726e33476eab758f2; please unresolve if not?

> * E2E tests with real use cases are incorporated into the testing infrastructure.

Reviewer 3: Here's another: https://github.com/cardano-foundation/CIPs/pull/524#discussion_r1255491481 🙏.

and will utilize the [cryptonite](https://github.com/haskell-crypto/cryptonite/blob/master/Crypto/Hash/Keccak.hs) implementation. 
Since `cryptonite` is already a part of the [`cardano-base`](https://github.com/input-output-hk/cardano-base/blob/master/cardano-crypto-class/src/Cardano/Crypto/Hash/Keccak256.hs), 
this simplifies its integration into Plutus. The cost of the `keccak_256` operation will scale linearly with the length of the message.

More specifically, Plutus will gain the following primitive operation:

* `keccak_256` :: ByteString -> ByteString

The input to this function can be a `ByteString` of arbitrary size, and the output will be a `ByteString` of 32 bytes. 
Note that this function aligns with the format of existing hash functions in Plutus, such as [blake2b_256](https://github.com/input-output-hk/plutus/blob/75267027f157f1312964e7126280920d1245c52d/plutus-core/plutus-core/src/Data/ByteString/Hash.hs#L25)

## Rationale: how does this CIP achieve its goals?
While the `keccak256` function might be implemented in on-chain scripts, doing so would be computationally unfeasible. 

> 

Reviewer 2: ```suggestion

The library, cryptonite, is not implemented by and under control of the Plutus team. However, 
* It is a library already used in the Cardano stack to expose SHA3, and can be considered as a trustworthy implementation. 
* The function does not throw any exceptions as hash functions are defined to work with any ByteString input. It does not expect a particular particular structure. 
* It's behaviour is predictable. As mentioned above, the cost of the function is linear with respect to the size of the message provided as input. This is the same behaviour that other hash functions exposed in plutus (blake, sha3) have.
```

## Path to Active
This CIP may transition to active status once the Plutus version containing the `keccak_256` function is introduced 
in a node release and becomes available on Mainnet.

### Acceptance Criteria
* A Plutus binding is created for the `keccak256` function and included in a new version of Plutus.
* Integration tests, similar to those of the existing Plutus hash functions, are added to the testing infrastructure.

> * The function is benchmarked to assess its cost.

Reviewer 2: ```suggestion
* The function is benchmarked to assess its cost. As for other hash functions available in Plutus (blake2b and sha256), we expect the cost of keccak to be linear with respect to the size of the message.
```

Reviewer 2: @rphair . I've suggested the comment that we expect the cost of the function to be linear wrt the message size as part of the AC. Do you think it should go somewhere else?

Reviewer: @iquerejeta it looks sufficient to me there.  @michaelpj I don't know if a more detailed cost analysis is required in the CIP itself as per cip35... I was struggling to hold onto any concrete requirements from the meeting so maybe @KtorZ @SebastienGllmt could better answer this.

Reviewer: Resolved in https://github.com/cardano-foundation/CIPs/pull/524/commits/b0dc15e3cc43c6486c3656ac621991aa9ea87d21.

> * The function is benchmarked to assess its cost. As for other hash functions available in Plutus (blake2b and sha256), we expect the cost of keccak to be linear with respect to the size of the message.

Reviewer 2: ```suggestion
* The function is benchmarked to assess its cost. As for other hash functions available in Plutus (blake2b and sha256), we expect the cost of keccak to be linear with respect to the size of the message.
* The ledger is updated to include new protocol parameters to control costing of the new builtins.
```
